accepted/file-enumeration.md

```diff
@@ -0,0 +1,238 @@
+
+Recursive enumeration is also problematic in that there is no way to handle error states such as access issues or cycles created by links.
+
+These restrictions have a significant impact on file system intensive applications, a key example being MSBuild. We can address these restrictions and provide performant, configurable enumeration.
+
+## Scenarios and User Experience
+
+1. MSBuild can custom filter filesystem entries with limited allocations and form the results in any desired format.
+2. Users can build custom enumerations utilizing completely custom or provided commonly used filters and transforms.
+
+## Requirements
+
+1. Filtering based on common file system data is possible
+	a. Name
+	b. Attributes
+	c. Time stamps
+    d. File size
+2. Result transforms can be of any type
+3. We provide common filters and transforms
+4. Recursive behavior is configurable
+5. Error handling is configurable
+
+### Goals
+
+1. API is cross platform generic
+2. API minimizes allocations while meeting #1
+
+### Non-Goals
+
+1. API will not expose platform specific data
+3. Error handling configuration is fully customizable
+
+## Design
+
+### Proposed API surface
+
+``` C#
+namespace System.IO
+{
+    /// <summary>
+    /// Delegate for filtering out find results.
+    /// </summary>
+    internal delegate bool FindPredicate<TState>(ref RawFindData findData, TState state);
+
+    /// <summary>
+    /// Delegate for transforming raw find data into a result.
+    /// </summary>
+    internal delegate TResult FindTransform<TResult, TState>(ref RawFindData findData, TState state);
+
+    [Flags]
+    public enum FindOptions
+	{
+        None = 0x0,
+
+        // Enumerate subdirectories
+        Recurse = 0x1,
+
+        // Skip files/directories when access is denied
+        IgnoreAccessDenied = 0x2,
+
+        // Future: Add flags for tracking cycles, etc. 
+	}
+
+    public class FindEnumerable<TResult, TState> : CriticalFinalizerObject, IEnumerable<TResult>, IEnumerator<TResult>
+    {
+        public FindEnumerable(
+            string directory,
+            FindTransform<TResult, TState> transform,
+            FindPredicate<TState> predicate,
+            // Only used if FindOptions.Recurse is set. Default is to always recurse.
+            FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None)
+    }
+
+    public static class Directory
+    {
+        public static IEnumerable<TResult> Enumerate<TResult, TState>(
+            string path,
+            FindTransform<TResult, TState> transform,
+            FindPredicate<TState> predicate,
+			FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None);
+    }
+
+    public static class DirectoryInfo
```

Comment:
DirectoryInfo is not static of course

Reply:
Yup, misstype

---
In reply to: [154807425](https://github.com/dotnet/designs/pull/24#discussion_r154807425) [](ancestors = 154807425)